### DISCUSSION TITLE
Triangulation

### DISCUSSION THREAD

**Discussion by mecejide** [SupposeNot/RAMP#186] [Ideas]
I wish there was an operation that returned the dual of the omnitruncation of a given polytope as a simplicial complex that simpcomp could then work with.

**Comment by CunningGabe** [comment]
I think this will do what you want for now:

```

Omnitruncation := function(M)
	local n;
	
	n := Rank(M);
	return Wythoffian([0..n-1], M);
	end;

FacetsInTermsOfVertices := function(M)
	local g, nv, verts, facets, L;
	
	g := ConnectionGroup(M);
	IsSggi(g); # Force a check
	
	verts := Orbits(VertexFigureSubgroup(g));
	nv := Size(verts);
	facets := Orbits(FacetSubgroup(g));
	
	L := List(facets, F -> Filtered([1..nv], i -> Size(Intersection(verts[i], F)) > 0));
	
	return L;
	end;

f := function(M)
	return SCFromFacets(FacetsInTermsOfVertices(Dual(Omnitruncation(M))));
	end;
```

Let me know if that works for you. Probably we will add those two functions above to core RAMP functionality.

**Reply by mecejide** [reply]
Uh-oh. Trying to load both RAMP and simpcomp gives an error:

Error, operation `Dual' was created as an attribute, use `DeclareAttribute' at /opt/gap-v4.12.2/.libs/../lib/oper.g:957 called from
<function "DeclareOperation">( <arguments> )
 called from read-eval loop at /opt/gap-v4.12.2/.libs/../pkg/RAMP-master/lib/duality.gd:7
you can 'quit;' to quit to outer loop, or
you can 'return;' to continue

**Comment by CunningGabe** [comment]
Hmm, that might be tough to fix gracefully. I suppose some namespace clashes between packages working on similar-ish problems is inevitable.

As a workaround for you for now, I guess you could try a multi-file search-and-replace, changing all instances of 'Dual(' to 'DualManiplex(' or something of the sort.

**Reply by SupposeNot** [reply]
I thought that the whole point of declaring the input types was to get around this problem. Do we need to be more specific in those function declarations? Did the simcomp people screw that up in their implementation? I'll try to poke at this this weekend, but I know I won't be able to dig into it until then.

**Reply by SupposeNot** [reply]
Alternately... do we need to make it an attribute as well?

**Reply by CunningGabe** [reply]
Yeah ok, that might fix it. I'm not sure off-hand if there is a particular reason we didn't do that at the start. It's probably more likely that we (really, I) didn't know why it mattered. But perhaps this shows why - maybe things (especially with short names) should be attributes by default, rather than operations?

The only thing we might want to be careful about at some point: Maniplexes cache all known values of attributes, so in principle if an attribute is a big object, then maybe we don't want to cache it. But it turns out we can turn off caching for a particular attribute using DisableAttributeValueStoring (https://docs.gap-system.org/doc/ref/chap13_mj.html#X7851E2DB79656DB0)

For now, I think we don't need to worry about that, so if you want to try converting Dual to an Attribute instead of an Operation, you're welcome to go ahead - I won't have a chance to deal with it for a while.

**Reply by SupposeNot** [reply]
I've been poking at this a little bit. I think we very much want the Dual to actually return an object, as opposed to being attached as an Attribute. I'm wondering if instead of using DeclareOperation we used DeclareConstructor? Or would we run into the same problem? Anyone know?

**Reply by SupposeNot** [reply]
Poking at it a little bit more... I guess it does no harm to have these be attributes? I mean, I can still set (for example) 
`g:=AutomorphismGroup(Cube(3))`
And I get access to the automorphism group... What could go wrong?

**Comment by SupposeNot** [comment]
Well, I've gone ahead and made the change to using Attribute for Dual. Seems to allow simpcomp to load without errors.
